### ISSUE TITLE
Console log output and events feature request

### ISSUE THREAD

**Issue by valtido** [whatwg/console#53]
Hi,

I would like to request for a couple features, on request we should have a method that can output everything console has recorded so far like a stack list.

This is useful when you need to save it programmatically rather than debugging manually.

I would also like an event listener, so we can assign a handler and do something when there is a new entry on log, warn, error etc. Again for the same reason.

**Comment by zspitzer** [comment]
like a `console.read(100);` where 100 is the max number of lines to return?

**Comment by valtido** [comment]
Yeah sure, but this doesn't work at the moment, and also could be `console.on("all,error,log,warning...", callback)`

Or promise instead of callback. :)

Cheers

**Comment by domfarolino** [comment]
Seems like a good idea. Some points:

1.) I think we definitely wouldn't want a promise for this sort of thing. Sounds like you'd like to extend `addEventListener` to console? Like `console.addEventListener('warn', e => {...})` kind of? I think that could be useful. Promises don't really fit this sort of behavior IMO.

2.) IIRC as @terinjokes pointed out to me the other day in person, callbacks should only be triggered for events originating from scripts whose [errors are not muted](https://html.spec.whatwg.org/multipage/webappapis.html#muted-errors) for security.

**Comment by terinjokes** [comment]
We've defined an interface in this specification. I think there's more benefit to being able to pass an implementation of that interface (which might be *this* Console, or something created at runtime) to the library, rather than adding and requiring eventing supporting to all console implementations.

**Comment by zspitzer** [comment]
would console.read() also include the script source? i.e.  "some message" app.js:45

**Comment by domfarolino** [comment]
I agree with @terinjokes on this, and I think there's more benefit to being able to build a mechanism that can do some sort of tracking when console methods are called as opposed to build that in here. This is mostly because it is only application code that calls the console methods, so application code authors (devs) should be able to easily track when these methods are being called.

It *may* be useful however, to be alerted when warnings are reported to the console (see #57), I guess it'd be worth polling implementers there, but I'm not sure.

Regarding console.read, there might be benefit to storing and having access to previous console output, but I think I'd need a more compelling use-case than https://github.com/w3c/web-platform-tests/issues/8654 (testing). As discussed a bit in https://github.com/whatwg/console/issues/75, there is some use but most of it isn't terribly productive so for now I'm kind of against it.

**Comment by Thoughtscript** [comment]
Per https://github.com/whatwg/console/issues/245 - there are some newer Browser functionalities that might benefit and be benefited by this kind of functionality now (that likely didn’t exist at the time the Issue was first created/opened).

There are now Read-Only Browser Warnings (in most browsers I’ve tested - specifics in the other Issue). These are often intended to be actionable and some kind of Console Field, Listener, etc. would allow these kinds of issues to be addressed programmatically.
